File: source/ch/cyberduck/ui/controller/CreateFileController.cs
// 
// Copyright (c) 2010-2013 Yves Langisch. All rights reserved.
// http://cyberduck.ch/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// yves@cyberduck.ch
// 

using System;
using System.Collections.ObjectModel;
using System.Windows.Forms;
using Ch.Cyberduck.Ui.Controller.Threading;
using ch.cyberduck.core;
using ch.cyberduck.core.editor;
using ch.cyberduck.core.features;
using UploadTargetFinder = ch.cyberduck.ui.browser.UploadTargetFinder;
using DisabledApplicationQuitCallback = ch.cyberduck.core.local.DisabledApplicationQuitCallback;
using DisabledTransferErrorCallback = ch.cyberduck.core.transfer.DisabledTransferErrorCallback;
using java.util;

namespace Ch.Cyberduck.Ui.Controller
{
    internal class CreateFileController : FileController
    {
        public CreateFileController(IPromptView view, BrowserController browserController)
            : base(view, browserController)
        {
        }

        public override void Callback(DialogResult result)
        {
            if (!String.IsNullOrEmpty(View.InputText) && !View.InputText.Trim().Equals(string.Empty))
            {
                Path parent = new UploadTargetFinder(Workdir).find(BrowserController.SelectedPath);
                if (DialogResult.OK == result)
                {
                    BrowserController.background(new CreateFileAction(BrowserController, parent, View.InputText, false));
                }
                if (DialogResult.Yes == result)
                {
                    BrowserController.background(new CreateFileAction(BrowserController, parent, View.InputText, true));
                }
            }
        }

        private class CreateFileAction : BrowserControllerBackgroundAction
        {
            private readonly bool _edit;
            private readonly Path _file;
            private readonly string _filename;
            private readonly Path _workdir;

            public CreateFileAction(BrowserController controller, Path workdir, string filename, bool edit)
                : base(controller)
            {
                _workdir = workdir;
                _filename = filename;
                _edit = edit;
                _file = new Path(_workdir, _filename, EnumSet.of(AbstractPath.Type.file));
            }

            public override object run()
            {
                Session session = BrowserController.Session;
                Touch feature = (Touch) session.getFeature(typeof (Touch));
                feature.touch(_file);
                if (_edit)
                {
                    _file.attributes().setSize(0L);
                    Editor editor = EditorFactory.instance().create(BrowserController, BrowserController.Session, _file);
                    editor.open(new DisabledApplicationQuitCallback(), new DisabledTransferErrorCallback());
                }
                return true;
            }

            public override void cleanup()
            {
                base.cleanup();
                if (_filename.StartsWith("."))
                {
                    BrowserController.ShowHiddenFiles = true;
                }
                BrowserController.RefreshParentPaths(new Collection<Path> {_file}, new Collection<Path> {_file});
            }

            public override string getActivity()
            {
                return String.Format(LocaleFactory.localizedString("Uploading {0}", "Status"), _file.getName());
            }
        }
    }
}

File: source/ch/cyberduck/ui/cocoa/CreateFileController.java
package ch.cyberduck.ui.cocoa;

/*
 *  Copyright (c) 2005 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.binding.application.NSAlert;
import ch.cyberduck.binding.application.NSImage;
import ch.cyberduck.core.Cache;
import ch.cyberduck.core.LocaleFactory;
import ch.cyberduck.core.Path;
import ch.cyberduck.core.Session;
import ch.cyberduck.core.editor.EditorFactory;
import ch.cyberduck.core.exception.BackgroundException;
import ch.cyberduck.core.features.Touch;
import ch.cyberduck.core.local.Application;
import ch.cyberduck.core.resources.IconCacheFactory;
import ch.cyberduck.ui.browser.UploadTargetFinder;
import ch.cyberduck.ui.cocoa.threading.BrowserControllerBackgroundAction;

import java.text.MessageFormat;
import java.util.Collections;
import java.util.EnumSet;

/**
 * @version $Id$
 */
public class CreateFileController extends FileController {

    public CreateFileController(final WindowController parent, final Cache<Path> cache) {
        super(parent, cache, NSAlert.alert(
                LocaleFactory.localizedString("Create new file", "File"),
                LocaleFactory.localizedString("Enter the name for the new file:", "File"),
                LocaleFactory.localizedString("Create", "File"),
                EditorFactory.instance().getDefaultEditor() != Application.notfound ? LocaleFactory.localizedString("Edit", "File") : null,
                LocaleFactory.localizedString("Cancel", "File")
        ));
        alert.setIcon(IconCacheFactory.<NSImage>get().documentIcon(null, 64));
    }

    @Override
    public void callback(final int returncode) {
        final Path parent = new UploadTargetFinder(this.getWorkdir()).find(this.getSelected());
        if(returncode == DEFAULT_OPTION) {
            this.run(parent, inputField.stringValue(), false);
        }
        else if(returncode == ALTERNATE_OPTION) {
            this.run(parent, inputField.stringValue(), true);
        }
    }

    protected void run(final Path parent, final String filename, final boolean edit) {
        final BrowserController c = (BrowserController) this.parent;
        final Path file = new Path(parent, filename, EnumSet.of(Path.Type.file));
        c.background(new BrowserControllerBackgroundAction<Path>(c) {
            @Override
            public Path run() throws BackgroundException {
                final Session<?> session = c.getSession();
                final Touch feature = session.getFeature(Touch.class);
                feature.touch(file);
                if(edit) {
                    file.attributes().setSize(0L);
                    c.edit(file);
                }
                return file;
            }

            @Override
            public String getActivity() {
                return MessageFormat.format(LocaleFactory.localizedString("Uploading {0}", "Status"),
                        file.getName());
            }

            @Override
            public void cleanup() {
                super.cleanup();
                if(filename.charAt(0) == '.') {
                    c.setShowHiddenFiles(true);
                }
                c.reload(Collections.singletonList(file), Collections.singletonList(file));
            }
        });
    }
}